File: Tatelier.Common/SongSelect/ScoreOverview.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading.Tasks;

namespace Tatelier.Common.SongSelect
{
	[DebuggerDisplay("Title={Title}")]
	public class ScoreOverview
		: IItem
	{
		/// <summary>
		/// ファイル種別
		/// </summary>
		enum FileType
		{
			/// <summary>
			/// Tatelierフォーマット譜面
			/// </summary>
			TatelierScore,
			/// <summary>
			/// TJA
			/// </summary>
			TJA,
			/// <summary>
			/// OSU
			/// </summary>
			OSU,
		}

		/// <summary>
		/// 
		/// </summary>
		/// <param name="scoreFolderPath">譜面フォルダパス</param>
		/// <param name="relativeFilePath">相対ファイルパス</param>
		/// <param name="categoryControl">カテゴリ管理クラス</param>
		/// <returns>インスタンス</returns>
		public static ScoreOverview Create(string scoreFolderPath, string relativeFilePath, CategoryControl categoryControl)
		{
			var extension = Path.GetExtension(relativeFilePath);

			FileType type;

			switch (extension)
			{
				case ".tja":
				case ".pts":
					type = FileType.TJA;
					break;
				case ".tlscore":
				default:
					type = FileType.TatelierScore;
					break;
			}

			var score = new ScoreOverview(scoreFolderPath, relativeFilePath, type, categoryControl);

			return score;
		}

		/// <summary>
		/// ファイルパス(相対パス)
		/// </summary>
		public string FilePath;

		/// <summary>
		/// タイトル
		/// </summary>
		public string Title;

		/// <summary>
		/// ID
		/// </summary>
		public string Id;

		/// <summary>
		/// カテゴリ名リスト
		/// </summary>
		public string[] Category;

		/// <summary>
		/// 音源
		/// </summary>
		public string Wave;

		/// <summary>
		/// デモ音源
		/// </summary>
		public string DemoMusicWave;

		/// <summary>
		/// デモ再生開始時間(秒)
		/// </summary>
		public double DemoStart;

		/// <summary>
		/// カテゴリデータ
		/// </summary>
		public Category[] CategoryData;

		public SongSelectItemType Type => SongSelectItemType.Score;

		public string RelativeFilePath => FilePath;

		/// <summary>
		/// TJA読込
		/// </summary>
		/// <param name="filePath">ファイルパス</param>
		/// <param name="categoryControl">カテゴリクラス</param>
		void LoadTJA(string filePath, CategoryControl categoryControl)
		{
			using (var sr = new StreamReader(filePath, Encoding.GetEncoding("shift-jis")))
			{
				var regex = new Regex(@"^(\S+):(.+)$");

				while (!sr.EndOfStream)
				{
					var line = sr.ReadLine();

					if (regex.IsMatch(line))
					{
						var match = regex.Match(line);
						var groups = match.Groups;

						switch (groups[1].Value.ToUpper())
						{
							case "TITLE":
								{
									Title = groups[2].Value;
								}
								break;
							case "WAVE":
								{
									Wave = groups[2].Value;
								}
								break;
							case "DEMOWAVE":
								{
									DemoMusicWave = groups[2].Value;
								}
								break;
							case "GENRE":
								{
									Category = groups[2].Value.Split(new string[] { "|" }, StringSplitOptions.RemoveEmptyEntries);
									CategoryData = categoryControl?.Get(Category.Select(v => v.Split(',')[0]).ToArray()) ?? new Category[0];
								}
								break;
						}
					}
				}

				// カテゴリデータがない場合はその他を取得する
				if (CategoryData == null) CategoryData = categoryControl.GetOther();

				for (int i = 0; i < CategoryData.Length; i++)
				{
					CategoryData[i]?.Add(this, "");
				}
			}
		}

		void LoadTatelierScore(string filePath, CategoryControl categoryControl)
		{

		}

		/// <summary>
		/// コンストラクタ
		/// </summary>
		/// <param name="scoreFolderPath">譜面フォルダパス</param>
		/// <param name="relativeFilePath">ファイル相対パス</param>
		/// <param name="fileType">ファイル種別</param>
		/// <param name="categoryControl">カテゴリ管理</param>
		ScoreOverview(string scoreFolderPath, string relativeFilePath, FileType fileType, CategoryControl categoryControl)
		{
			string filePath = Path.Combine(scoreFolderPath, relativeFilePath);

			FilePath = relativeFilePath;

			switch (fileType)
			{
				case FileType.TatelierScore:
					LoadTatelierScore(filePath, categoryControl);
					break;
				case FileType.TJA:
					LoadTJA(filePath, categoryControl);
					break;
				case FileType.OSU:
					break;
			}

		}
	}
}

File: Tatelier.Nucleus/SongSelect/ScoreOverview.cpp
#include "ScoreOverview.h"

#include <fstream>
#include <functional>
#include <iostream>
#include <memory>
#include <sstream>

#include "../Utility.h"

#define ARRAY_SIZE(X) (sizeof(X) / (sizeof(X[0])))

namespace SongSelect {
	void ScoreOverview::LoadTJA(const char* filePath, std::shared_ptr<CategoryControl> categoryControl)
	{

		struct K {
			std::string Name;
			std::function<void(const std::string& text)> Func;
		};
		K k[] = {
			{ "TITLE:", [this](const std::string& text) { this->title = ttleConvertToSjis(text); } },
			{ "WAVE:", [this](const std::string& text) { this->wave = ttleConvertToSjis(text); } },
			{ "GENRE:", [this, categoryControl](const std::string& text) {
				 std::stringstream ss { text };
				 std::string buf;
				 while (std::getline(ss, buf, '|')) {
					 this->categories.push_back(ttleConvertToSjis(buf));
				 }
				 this->categoryData = categoryControl->Get(this->categories);
			 } },
			{ "DEMOWAVE:", [this](const std::string& text) { this->demoMusicWave = ttleConvertToSjis(text); } },
		};

		std::ifstream ifs(filePath);

		//
		if (!ifs) {
			return;
		}

		std::string buf;
		while (!ifs.eof()) {
			if (std::getline(ifs, buf)) {
				for (int i = 0; i < ARRAY_SIZE(k); i++) {
					if (ttleStartWith(buf.c_str(), k[i].Name.c_str())) {
						k[i].Func(buf.replace(buf.find(k[i].Name), k[i].Name.size(), ""));
					}
				}
			}
		}

		if (categoryData.size() == 0) {
			categoryData = categoryControl->GetOther();
		}
		auto selectItem = shared_from_this();
		for (int i = 0; i < categoryData.size(); i++) {
			std::shared_ptr<ScoreOverview> p = std::dynamic_pointer_cast<ScoreOverview>(selectItem);
			categoryData[i]->Add(std::static_pointer_cast<Item>(p), "");
		}
	}
	ScoreOverview::ScoreOverview(const std::filesystem::path& scoreFolderPath, const std::filesystem::path& relativeFilePath, FileType fileType, std::shared_ptr<CategoryControl> categoryControl)
	{
		std::filesystem::path fp = scoreFolderPath / relativeFilePath;
		filePath = relativeFilePath.string();

		auto selectItem = std::shared_ptr<ScoreOverview>(this);

		switch (fileType) {
		case FileType::TatelierScore:
			break;
		case FileType::TJA:
			LoadTJA(filePath.c_str(), categoryControl);
			break;
		case FileType::OSU:
			break;
		}
	}
	const std::string& ScoreOverview::GetTitle()
	{
		return title;
	}
	ScoreOverview* ScoreOverview::CreateAndRegist(const std::filesystem::path& scoreFolderPath, const std::filesystem::path& relativeFilePath, std::shared_ptr<CategoryControl> categoryControl)
	{

		FileType type;
		switch (ttleEndWithIndex(relativeFilePath.string(), ".tja|.tlscore")) {
		case 1:
			type = FileType::TJA;
			break;
		case 2:
			type = FileType::TatelierScore;
			break;
		default:
			return nullptr;
		}

		return new ScoreOverview(scoreFolderPath, relativeFilePath, type, categoryControl);
	}
	const SongSelectItemType& ScoreOverview::GetType()
	{
		return SongSelectItemType::Score;
	}
}